### ISSUE TITLE
L043 doesn't consider THEN statement in evaluation, alters selection logic

### ISSUE THREAD

**Issue by dannyshaw** [sqlfluff/sqlfluff#2903]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When fixing in rule L043
```sql
SELECT
    COUNT(DISTINCT   
        CASE
            WHEN item.submitted_timestamp IS NOT NULL
            THEN item.sitting_id
        END
    ) AS num_submitted
FROM table
```
is modified to :
```sql
SELECT
    COUNT(DISTINCT
        response.submitted_timestamp
    ) AS num_submitted
FROM table
```

This is a different expression and means any submissions with the same timestamp will not be considered as distinct counts.

### Expected Behaviour

The `THEN` statement should probably be inspected to see if it's a different expression to the NOT NULL check in the WHEN.

### Observed Behaviour

If the THEN statement isn't inspected as part of the rule, this is going to regularly be changing the logic of CASE statements.

### How to reproduce

As above.

### Dialect

postgres

### Version

0.11.1

### Configuration

```
[sqlfluff]
dialect = postgres
templater = dbt

# Excluded rules (see https://docs.sqlfluff.com/en/stable/rules.html):
# L009: File end with newline - suggested in the docs when using the DBT templater
# L016: Line too long - bugged, doesn't honour line lengths in lines containing templating syntax
# L025: No unused table aliases - bugged, doesn't consider table aliases used in join criteria or subqueries, which can break queries
# L031: No table aliases in joins or froms - tables aliases are required in DBT if you're using ref or src
# L034: Column ordering - I think query authors should have explicit control over the order of columns
# L042: No subqueries in joins or froms - I think query authors should be able to use subqueries in JOIN clauses
exclude_rules = L009,L016,L025,L031,L034,L042

[sqlfluff:rules]
tab_space_size = 2

[sqlfluff:templater:dbt]
target = dev
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by barrywhart** [comment]
Ooh, nasty!

**Comment by barrywhart** [comment]
Simplified test case:
```
SELECT
    CASE
        WHEN item.submitted_timestamp IS NOT NULL
        THEN item.sitting_id
    END
```
